README.md
R package ‘handyFunctions’ version 0.1.0
================
Copyright (C) 2022, Hongfei Liu
2022/08/16

<!-- README.md is generated from README.Rmd. Please edit that file -->

# handyFunctions

<!-- badges: start -->

![](https://lifecycle.r-lib.org/articles/figures/lifecycle-experimental.svg)
[![Codecov test
coverage](https://codecov.io/gh/LuffyLouis/handyFunctions/branch/main/graph/badge.svg)](https://app.codecov.io/gh/LuffyLouis/handyFunctions?branch=main)
[![R-CMD-check](https://github.com/LuffyLouis/handyFunctions/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/LuffyLouis/handyFunctions/actions/workflows/R-CMD-check.yaml)

<!-- badges: end -->

The goal of handyFunctions is to get rid of the barrier to deal with
non-standard data.frame format for R newbies, especially the user in
bioinformatics data analysis. Besides, there are also some required plot
functions for downstream analysis of dataset generated from vcftools and
plink.

## Table of contents

-   [Installation](#installation)
-   [Example](#example)
    -   [ReformatDataframe](#reformatdataframe)
    -   [InteractDataframe](#interactdataframe)
    -   [Post-VCF](#post-vcf)
        -   [plotSNVdensity](#plotsnvdensity)
-   [License](#license)

## Installation

You can install the development version of handyFunctions like so:

``` r
## Clone it from github and install it locally
git clone https://github.com/LuffyLouis/handyFunctions.git
## OR
## Install it in R
remotes::install_github("LuffyLouis/handyFunctions")
```

## Example

`handyFunctions`package contain three main sections, including
ReformatDataframe, InteractDataframe, and Post-VCF. There are some basic
examples which show you how to solve common problems in data analysis:

### ReformatDataframe

This section is designed to reformat data.frame with odd colnames,
rownames, or even inappropriate dtypes for each columns.

#### unifyDataframe

Based on the following example `unifyDataframe` function, you can change
the formats of raw data.frame to what you want. Especially for the
dtypes in data.frame, you can set `custom=FALSE` for automatically
changing into appropriate dtypes.

``` r
library(handyFunctions)
data("people")
head(people)
#>          ..name  ..sex ..age ..death..age
#> 1       Ming Li   male    12           34
#> 2    Zixuan Liu female    23       thirty
#> 3    Yizhen Zhu   male    NA           54
#> 4 Lingling Wang female    21           77
#> 5      Bang Wei   male    11         <NA>
#> 6   Xiaoyu Chen female    74           89
modifiedPeople <- unifyDataframe(people,rawColSep = '[.][.]')
head(modifiedPeople)
#>   name    sex age death_age
#> 1    3   male  12        34
#> 2    6 female  23    thirty
#> 3    5   male  NA        54
#> 4    2 female  21        77
#> 5    1   male  11      <NA>
#> 6    4 female  74        89
```

**Note**: due to the separation supporting regEx, please use the
`"[.][.]"` for reformatting `people` data.frame.

### InteractDataframe

The `InteractDataframe` section is designed for interaction between two
data.frame or one data.frame and another vector.

#### mergeCustom

Sometimes, we often find it fuzzy and tedious while we’d like to merge
two data.frame with different colnames. Therefore, `mergeCustom`
function may be the better solution to get it rid of.

``` r
library(handyFunctions)
data("people");data("grade")
head(people)
#>          ..name  ..sex ..age ..death..age
#> 1       Ming Li   male    12           34
#> 2    Zixuan Liu female    23       thirty
#> 3    Yizhen Zhu   male    NA           54
#> 4 Lingling Wang female    21           77
#> 5      Bang Wei   male    11         <NA>
#> 6   Xiaoyu Chen female    74           89
head(grade)
#>            name chinese math english physics biology chemistry
#> 1       Ming Li     120  130     145      80      90        99
#> 2    Zixuan Liu     109  120     110      85      99        88
#> 3    Yizhen Zhu      98  113     100      74     100        76
#> 4 Lingling Wang     138  145     126      55      89       100
#> 5      Bang Wei     119  105     139     100      78        99
#> 6   Xiaoyu Chen     119  105     120      69      80        77
merged <- mergeCustom(people,grade,xcol = '..name',ycol = 'name')
head(merged)
#>          ..name  ..sex ..age ..death..age chinese math english physics biology
#> 1      Bang Wei   male    11         <NA>     119  105     139     100      78
#> 2 Lingling Wang female    21           77     138  145     126      55      89
#> 3       Ming Li   male    12           34     120  130     145      80      90
#> 4   Xiaoyu Chen female    74           89     119  105     120      69      80
#> 5    Yizhen Zhu   male    NA           54      98  113     100      74     100
#> 6    Zixuan Liu female    23       thirty     109  120     110      85      99
#>   chemistry
#> 1        99
#> 2       100
#> 3        99
#> 4        77
#> 5        76
#> 6        88
```

### Post-VCF

#### plotSNVdensity

``` r
library(handyFunctions)
library(ggplot2)
data("SNV_1MB_density_data")
head(SNV_1MB_density_data)
#>   CHROM BIN_START SNP_COUNT VARIANTS.KB
#> 1     1         0       253       0.253
#> 2     1   1000000        31       0.031
#> 3     1   2000000       208       0.208
#> 4     1   3000000        77       0.077
#> 5     1   4000000       204       0.204
#> 6     1   5000000        75       0.075
ShowSNPDensityPlot(SNV_1MB_density_data,binSize=1e6,chromSet = c(38:1),withchr=T)+
  theme(axis.text.y = element_text(size=12))
#> ## Filtering the density data with specific chrom set...
#> ## Judging if it should be added the chr and factoring...
#> ## Reformatting the raw density data...
```

<img src="man/figures/README-plotSNVdensity-1.png" width="100%" />

## License

This project is published under MIT license, see file `LICENSE`.


NEWS.md
# handyFunctions 0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/ReformatDataframe.R

#' Return reformatted data.frame with standard col names
#'
#' @param rawDataFrame Raw data.frame input
#' @param cols Specific col names or indexes what you want to reformat (default: TRUE, use all cols)
#' @param rawSep Raw odd separation symbol in col names of raw data.frame.
#' Note: it supports regEx (regular expression), so "." means all possible symbols.
#' If you want to use the "." dot notation, please use "[.]".
#'
#' @param sep Separation symbol in col names of modified data.frame
#'
#' @return A modified data.frame with col names separated by your given delimitator
#' @export
#'
#' @examples
#' library(handyFunctions)
#' data(people)
#' modified_people <- modifyColNames(people,rawSep = "[.][.]")
#'
modifyColNames <- function(rawDataFrame, cols = TRUE, rawSep = "..", sep = "_") {
  # rawDataFrame <<- rawDataFrame
  modifiedDataFrame <- rawDataFrame
  rawColNames <- colnames(rawDataFrame)
  rawColNamesIndex <- checkCols(rawDataFrame, cols)
  if (rawColNamesIndex[1] == FALSE) {
    return(FALSE)
  }
  rawColNamesModified <- rawColNames[rawColNamesIndex]
  # print(rawColNames)
  modifiedColNames <- unlist(lapply(rawColNamesModified, function(x) {
    temp <- stringr::str_trim(stringr::str_replace_all(x, rawSep, " "))
    # temp=strsplit(x,rawSep,fixed = T)[[1]];
    return(stringr::str_replace_all(temp, " ", sep))
  }))
  # print(modifiedColNames)

  colnames(modifiedDataFrame)[rawColNamesIndex] <- modifiedColNames
  # print(modifiedDataFrame)
  return(modifiedDataFrame)
}

#' check the validation and return index of cols given from input in rawDataFrame
#'
#' @param rawDataFrame raw data.frame
#' @param cols specific cols given from input
#'
#' @return return validation (only FALSE if invaild cols input) or index of cols
#' @export
#'
#' @examples
#' library(handyFunctions)
#' data(people)
#' checkCols(people, c("..name", "..sex"))
#' # OR
#' checkCols(people, c(1, 2))
checkCols <- function(rawDataFrame, cols) {
  rawColNames <- colnames(rawDataFrame)
  if (class(cols) == "character" | class(cols) == "numeric" & length(cols) > 1) {
    if (class(cols) == "character") {
      ColNamesIndex <- unlist(lapply(cols, function(x) {
        which(rawColNames == x)
      }))
    } else {
      ColNamesIndex <- cols
    }
  } else if (cols == T | cols == "TRUE" | cols == "T") {
    ColNamesIndex <- 1:length(colnames(rawDataFrame))
  } else {
    message("Error, cols is not legal!")
    return(FALSE)
  }
  return(ColNamesIndex)
}

#' Return suggested dtype of vector input
#'
#' @param vector vector/list input
#'
#' @return Return suggested dtypes of vector
#' @export
#'
#' @examples
#' library(handyFunctions)
#' vector <- c(1, 2, 3, "", NA, "  ", "four", "NA", 5)
#' checkDtype(vector)
#'
checkDtype <- function(vector) {
  # vector = temp[,1]
  # validVector = unlist(vector)
  # N = length(vector)
  # vector = c('sd ds','sds sd','ad sd')
  vector <- c(1, 2, 3, "", NA, "  ", "four", "NA", 5)
  vector <- as.character(unlist(vector))
  check_sum <- sum(unlist(lapply(vector, function(x) {
    !is.na(as.numeric(x))
  })), na.rm = T)
  check_nonsum <- sum(unlist(lapply(vector, function(x) {
    is.na(as.numeric(x))
  })), na.rm = T)
  check_blank <- sum(unlist(lapply(vector, function(x) {
    !is.na(stringr::str_match(x, "^ +$"))
  })), na.rm = T)
  check_null <- sum(unlist(lapply(vector, function(x) {
    x == ""
  })), na.rm = T)
  check_possNA <- sum(unlist(lapply(vector, function(x) {
    x == "NA"
  })), na.rm = T)
  check_NA <- sum(unlist(lapply(vector, function(x) {
    is.na(x)
  })), na.rm = T)

  N_num <- check_sum
  N_char <- check_nonsum - check_NA - check_possNA - check_null - check_blank
  if (N_char > 0) {
    possDtype <- "char"
  } else if (N_char == 0) {
    possDtype <- "num"
  }
  # if (max(N_num,N_char)<N){
  #   poss = FALSE
  # } else {
  #   poss = TRUE
  # }
  return(possDtype)
}


#' Return suggested appropriate dtypes for each column in rawDataFrame
#'
#' @param rawDataFrame Raw data.frame
#' @param cols Specify cols which you want to change its dtypes when custom is FALSE (default: TRUE, for all cols)
#' @param dtype Specify indexed matched dtypes whcih you want to update when custom is FALSE (default: FALSE, for automatically update)
#' @param custom Option whether set to auto/custom , you can specify your custom dtypes for cols given when setting to TRUE (default: FALSE, for auto)
#'
#' @return Return a new data.frame with appropriate dtypes suggested for each cols
#' @export
#'
#' @examples
#' library(handyFunctions)
#' data(people)
#' modifyColTypes(people)
#'
modifyColTypes <- function(rawDataFrame, cols = TRUE, dtype = FALSE, custom = FALSE) {
  # get col names
  allColNames <- colnames(rawDataFrame)
  ColNamesIndex <- checkCols(rawDataFrame, cols)
  if (ColNamesIndex[1] == FALSE) {
    return(FALSE)
  }
  # ColNamesIndex =
  if (custom == T) {
    if (dtype == FALSE) {
      message("Error, dtype is FALSE:\n You must specify the data type of the given cols!")
      return(FALSE)
    } else if (class(dtype) == "character" & length(dtype) >= 1) {
      if (length(dtype) == length(cols)) {
        for (i in 1:length(ColNamesIndex)) {
          if (!is.na(stringr::str_match(dtype[i], "num"))) {
            rawDataFrame[, ColNamesIndex[i]] <- as.numeric(rawDataFrame[, ColNamesIndex[i]])
          } else if (!is.na(stringr::str_match(dtype[i], "char"))) {
            rawDataFrame[, ColNamesIndex[i]] <- as.character(rawDataFrame[, ColNamesIndex[i]])
          } else {
            message(paste0("Error,", i, "-indexed value in dtype is not legal, which should contain num or char!"))
            return(FALSE)
          }
        }
      } else {
        message("Error, dtype is not match with cols:\n The indexes of dtype must be matched with that of cols!")
        return(FALSE)
      }
    }
  }
  if (custom == F) {
    for (i in 1:length(ColNamesIndex)) {
      if (checkDtype(rawDataFrame[, i]) == "num") {
        rawDataFrame[, ColNamesIndex[i]] <- as.numeric(rawDataFrame[, ColNamesIndex[i]])
      } else if (checkDtype(rawDataFrame[, i]) == "char") {
        rawDataFrame[, ColNamesIndex[i]] <- as.character(rawDataFrame[, ColNamesIndex[i]])
      } else {
        warning(paste0("Warning, the ", i, "-indexed col (", allColNames[ColNamesIndex[i]], ") cannot be detected as approriate dtype!"))
        return(FALSE)
      }
    }
  }

  return(rawDataFrame)
}

#' Return reformatted data.frame with standard row names
#'
#' @param rawDataFrame Raw data.frame input
#' @param rows Specific row names or indexes what you want to reformat (default: TRUE, use all row)
#' @param rawSep Raw odd separation symbol in row names of raw data.frame.
#' Note: it supports regEx (regular expression), so "." means all possible symbols.
#' If you want to use the "." dot notation, please use "[.]".
#' @param sep Separation symbol in row names of modified data.frame
#'
#' @return A modified data.frame with row names separated by your given delimitator
#' @export
#'
#' @examples
#' library(handyFunctions)
#' data(people)
#' modifyRowNames(people)
#'
modifyRowNames <- function(rawDataFrame, rows = TRUE, rawSep = "..", sep = "_") {
  # rawDataFrame <<- rawDataFrame
  modifiedDataFrame <- rawDataFrame
  rawRowNames <- rownames(rawDataFrame)
  if (class(rows) == "character" | class(rows) == "numeric" & length(rows) > 1) {
    rawRowNamesIndex <- unlist(lapply(rows, function(x) {
      which(rawRowNames == x)
    }))
  } else if (rows == T | rows == "TRUE" | rows == "T") {
    rawRowNamesIndex <- 1:length(rownames(rawDataFrame))
  }
  rawRowNamesModified <- rawRowNames[rawRowNamesIndex]
  modifiedRowNames <- unlist(lapply(rawRowNamesModified, function(x) {
    temp <- stringr::str_trim(stringr::str_replace_all(x, rawSep, " "))
    # temp=strsplit(x,rawSep,fixed = T)[[1]];
    return(stringr::str_replace_all(temp, " ", sep))
  }))

  rownames(modifiedDataFrame)[rawRowNamesIndex] <- modifiedRowNames
  # print(modifiedDataFrame)
  return(modifiedDataFrame)
}


#' Reformat dataframe with the all modifiers simultaneously (colNames, rowNames and dtypes)
#'
#' @param rawDataFrame raw data.frame
#' @param rawRowSep raw separation deliminator of row names in raw data.frame
#' @param rowSep the new separation deliminator of row names
#' @param rawColSep raw separation deliminator of col names in raw data.frame
#' @param colSep the new separation deliminator of col names
#' @param changeDtype if change the dtypes of cols
#'
#' @return A modified data.frame with applied to above all modifiers
#' @export
#'
#' @examples
#' library(handyFunctions)
#' data(people)
#' unifyDataframe(people,rawColSep = "[.][.]")
#'
unifyDataframe <- function(rawDataFrame, rawRowSep = "..", rowSep = "_", rawColSep = "..", colSep = "_", changeDtype = TRUE) {
  modifiedDataFrame <- modifyColNames(rawDataFrame, cols = T, rawSep = rawColSep, sep = colSep)
  modifiedDataFrame <- modifyRowNames(modifiedDataFrame, rows = T, rawSep = rawColSep, sep = colSep)
  if (changeDtype == T) {
    modifiedDataFrame <- modifyColTypes(modifiedDataFrame)
  }
  return(modifiedDataFrame)
}


#' Return specific-indexed vector according to given delimitator/separator by splitting one col in data.frame
#'
#' @param data vector or data.frame input
#' @param col the col names or indexes if data.frame input
#' @param sep separation deliminator
#' @param index the index of symbol which you want
#' @param fixed logical. If TRUE match split exactly, otherwise use regular expressions, detailed info can be seen in \href{/library/base/help/strsplit}{strsplit}.
#'
#' @return specific-indexed vector or factor
#' @export
#'
#' @examples
#' library(handyFunctions)
#' data(people)
#' splitCol(people, col = 1, sep = " ", index = 2)
#'
splitCol <- function(data, col = FALSE, sep, index, fixed = TRUE) {
  level <- NULL
  isFactor <- FALSE

  if (class(data) == "data.frame") {
    if (col == FALSE) {
      message("Error: please specify the col!")
      return(FALSE)
    } else {
      inputVector <- data[, col]
    }
  } else if (class(data) == "character") {
    inputVector <- data
  } else if (class(data) == "factor") {
    # warning("The class of ",substitute(data), 'is factor!\n It has been changed into character dtype!')
    # inputVector = as.character(data)
    inputVector <- data
  } else {
    inputVector <- NULL
    message(paste0("Error: ", substitute(data), " is NULL!"))
    return(FALSE)
  }

  if (class(inputVector) == "factor") {
    isFactor <- TRUE
    level <- levels(inputVector)
    inputVector <- as.character(inputVector)
  }

  outputVector <- unlist(lapply(inputVector, function(x) {
    strsplit(x, sep, fixed)[[1]][index]
  }))
  if (isFactor == T) {
    return(factor(outputVector, levels = level))
  } else {
    return(outputVector)
  }
}


..Rcheck/handyFunctions-Ex.R
pkgname <- "handyFunctions"
source(file.path(R.home("share"), "R", "examples-header.R"))
options(warn = 1)
base::assign(".ExTimings", "handyFunctions-Ex.timings", pos = 'CheckExEnv')
base::cat("name\tuser\tsystem\telapsed\n", file=base::get(".ExTimings", pos = 'CheckExEnv'))
base::assign(".format_ptime",
function(x) {
  if(!is.na(x[4L])) x[1L] <- x[1L] + x[4L]
  if(!is.na(x[5L])) x[2L] <- x[2L] + x[5L]
  options(OutDec = '.')
  format(x[1L:3L], digits = 7L)
},
pos = 'CheckExEnv')

### * </HEADER>
library('handyFunctions')

base::assign(".oldSearch", base::search(), pos = 'CheckExEnv')
base::assign(".old_wd", base::getwd(), pos = 'CheckExEnv')
cleanEx()
nameEx("ShowSNPDensityPlot")
### * ShowSNPDensityPlot

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: ShowSNPDensityPlot
### Title: Function of showing SNP density at chromosome level
### Aliases: ShowSNPDensityPlot

### ** Examples

library(handyFunctions)
data(SNV_1MB_density_data)
ShowSNPDensityPlot(SNV_1MB_density_data, binSize = 1e6, chromSet = c(38:1))




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("ShowSNPDensityPlot", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("checkCols")
### * checkCols

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: checkCols
### Title: check the validation and return index of cols given from input
###   in rawDataFrame
### Aliases: checkCols

### ** Examples

library(handyFunctions)
data(people)
checkCols(people, c("..name", "..sex"))
# OR
checkCols(people, c(1, 2))



base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("checkCols", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("checkDtype")
### * checkDtype

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: checkDtype
### Title: Return suggested dtype of vector input
### Aliases: checkDtype

### ** Examples

library(handyFunctions)
vector <- c(1, 2, 3, "", NA, "  ", "four", "NA", 5)
checkDtype(vector)




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("checkDtype", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("matchIndex")
### * matchIndex

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: matchIndex
### Title: Return the index of source vector matched with query vector
### Aliases: matchIndex

### ** Examples

library(handyFunctions)
data(grade)
matchIndex(grade[, "name"], c("Ming Li", "Bang Wei"))




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("matchIndex", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("mergeCustom")
### * mergeCustom

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: mergeCustom
### Title: merge two data.frame based on xcol and ycol
### Aliases: mergeCustom

### ** Examples

library(handyFunctions)
data(people)
data(grade)
mergeCustom(people, grade, "..name", "name")




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("mergeCustom", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("modifyColNames")
### * modifyColNames

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: modifyColNames
### Title: Return reformatted data.frame with standard col names
### Aliases: modifyColNames

### ** Examples

library(handyFunctions)
data(people)
modified_people <- modifyColNames(people,rawSep = "[.][.]")




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("modifyColNames", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("modifyColTypes")
### * modifyColTypes

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: modifyColTypes
### Title: Return suggested appropriate dtypes for each column in
###   rawDataFrame
### Aliases: modifyColTypes

### ** Examples

library(handyFunctions)
data(people)
modifyColTypes(people)




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("modifyColTypes", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("modifyRowNames")
### * modifyRowNames

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: modifyRowNames
### Title: Return reformatted data.frame with standard row names
### Aliases: modifyRowNames

### ** Examples

library(handyFunctions)
data(people)
modifyRowNames(people)




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("modifyRowNames", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("queryingInfo")
### * queryingInfo

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: queryingInfo
### Title: return index of x data.frame with the given vector/list or ycol
###   in data.frame (if set the accurate match or not)
### Aliases: queryingInfo

### ** Examples

library(handyFunctions)
data(grade)
queryingInfo(grade, "name", "chinese", c("Ming Li", "Bang Wei"))




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("queryingInfo", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("splitCol")
### * splitCol

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: splitCol
### Title: Return specific-indexed vector according to given
###   delimitator/separator by splitting one col in data.frame
### Aliases: splitCol

### ** Examples

library(handyFunctions)
data(people)
splitCol(people, col = 1, sep = " ", index = 2)




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("splitCol", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("unifyDataframe")
### * unifyDataframe

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: unifyDataframe
### Title: Reformat dataframe with the all modifiers simultaneously
###   (colNames, rowNames and dtypes)
### Aliases: unifyDataframe

### ** Examples

library(handyFunctions)
data(people)
unifyDataframe(people,rawColSep = "[.][.]")




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("unifyDataframe", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
### * <FOOTER>
###
cleanEx()
options(digits = 7L)
base::cat("Time elapsed: ", proc.time() - base::get("ptime", pos = 'CheckExEnv'),"\n")
grDevices::dev.off()
###
### Local variables: ***
### mode: outline-minor ***
### outline-regexp: "\\(> \\)?### [*]+" ***
### End: ***
quit('no')


R/postVCF.R


#' Function of showing SNP density at chromosome level
#'
#' @param densityData the raw density data generated from vcftools
#' @param binSize the bin size set while generating density data
#' @param densityColorBar vector Specific the color bar for plotting density plot (generally four colors)
#' @param chromSet vector Filtered chrom set which you want to plot (it must be matched with the CHROM column in densityData)
#' @param withchr logical If the chromsome labels of density plot is prefixed with "chr".
#' Note: it cannot work when the filtered chrom set contain other uncommon chrom symbols (e.g. NC0*, etc)
#'
#' @return A ggplot2 object for SNP density plot
#' @export
#'
#' @import ggplot2
#' @importFrom stats reorder quantile
#'
#' @examples
#' library(handyFunctions)
#' data(SNV_1MB_density_data)
#' ShowSNPDensityPlot(SNV_1MB_density_data, binSize = 1e6, chromSet = c(38:1))
#'
ShowSNPDensityPlot <- function(densityData, binSize, densityColorBar = c("grey", "darkgreen", "yellow", "red"),
                               chromSet = c(1:22), withchr = FALSE) {
  "
  densityData should be like this from vcftools:
    CHROM BIN_START SNP_COUNT VARIANTS.KB
  1     1    240000         1         0.1
  2     1    250000         0         0.0
  3     1    260000         0         0.0
  ...
  "
  # densityData<-SNV_density_data
  message("## Filtering the density data with specific chrom set...")
  index <- matchIndex(densityData[, "CHROM"], chromSet, T)
  tempSNPDensityData <- densityData[index, ]
  rawChrSet <- unique(tempSNPDensityData[, "CHROM"])
  message("## Judging if it should be added the chr and factoring...")
  isChr <- sum(!is.na(stringr::str_match(rawChrSet, "[C,c][H,h][R,r]")))
  isNum <- sum(!is.na(stringr::str_match(rawChrSet, "\\d+")))
  if (withchr == T) {
    if (isChr == length(rawChrSet)) {
      message("## Raw chr set is chr set, so it won't be changed!")
      tempSNPDensityData[, "CHROM"] <- factor(tempSNPDensityData[, "CHROM"], levels = chromSet)
    } else if (isChr == 0 & isNum == length(rawChrSet)) {
      tempSNPDensityData[, "CHROM"] <- paste0("chr", tempSNPDensityData[, "CHROM"])
      tempSNPDensityData[, "CHROM"] <- factor(tempSNPDensityData[, "CHROM"], levels = paste0("chr", chromSet))
    } else {
      warning("There are other chromosome symbols, so it won't be changed, or you can remove or replace that!!")
    }
  } else {
    if (isChr == length(rawChrSet)) {
      tempSNPDensityData[, "CHROM"] <- stringr::str_replace_all(tempSNPDensityData[, "CHROM"], "[C,c][H,h][R,r]", "")

      tempSNPDensityData[, "CHROM"] <- factor(tempSNPDensityData[, "CHROM"], levels = paste0(
        "chr",
        stringr::str_replace_all(chromSet, "[C,c][H,h][R,r]", "")
      ))
    } else if (isChr == 0 & isNum == length(rawChrSet)) {
      message("## Raw chr set is numbered set, so it won't be changed!")
      tempSNPDensityData[, "CHROM"] <- factor(tempSNPDensityData[, "CHROM"], levels = chromSet)
    } else {
      warning("There are other chromosome symbols, so it won't be changed, or you can remove or replace that!")
    }
  }


  ## reformat data
  message("## Reformatting the raw density data...")
  tempSNPDensityData[, "LEN"] <- binSize
  chrSet <- unique(tempSNPDensityData[, "CHROM"])
  i <- 1
  for (chrom in chrSet) {
    formerStart <- min(tempSNPDensityData$BIN_START[tempSNPDensityData[, "CHROM"] == chrom])
    temp <- data.frame("CHROM" = chrom, "BIN_START" = 0, "SNP_COUNT" = 0, "VARIANTS.KB" = 0, "LEN" = formerStart)
    if (i == 1) {
      formerStartData <- temp
    } else {
      formerStartData <- rbind(formerStartData, temp)
    }
    i <- i + 1
  }
  tempSNPDensityData <- rbind(tempSNPDensityData, formerStartData)
  minInColorBar <- 0
  # firstInColorBar<-quantile(tempSNPDensityData$SNP_COUNT,0.01)
  # secondInColorBar<-quantile(tempSNPDensityData$SNP_COUNT,0.5)
  thirdInColorBar <- quantile(tempSNPDensityData[, "SNP_COUNT"], 0.99)
  # maxInColorBar<-max(tempSNPDensityData$SNP_COUNT)

  p <- ggplot(tempSNPDensityData) +
    geom_bar(aes(x = LEN / 1e6, y = reorder(CHROM, CHROM), fill = SNP_COUNT), stat = "identity", position = "stack") +
    xlab("Position (Mb)") +
    scale_fill_gradientn(
      limits = c(minInColorBar, thirdInColorBar),
      # breaks = c(minInColorBar, firstInColorBar, secondInColorBar, thirdInColorBar),
      colours = densityColorBar
    ) +
    theme_classic() +
    theme(
      axis.title.x = element_text(size = 25),
      axis.title.y = element_blank(),
      axis.ticks.y = element_blank(),
      axis.text.x = element_text(size = 20),
      axis.text.y = element_text(size = 16),
      axis.line.y = element_blank()
    )
  return(p)
}
